### ISSUE TITLE
Allow external beautifiers for markup template tags

### ISSUE THREAD

**Issue by prettydiff** [prettydiff/prettydiff#323]
Extend markuppretty.js library to allow for external beautifiers on template types.

**Comment by lenilsondc** [comment]
Well done, I think this way other developers can contribute with more adapters from other languages.

By the way, Isn't it similar to jspretty, csvpretty and csspretty that already exists?

**Comment by prettydiff** [comment]
Yes.  jspretty, csspretty, and markuppretty are all written with a similar architectures.  csvpretty is different though.  I could pass the ASP code to jspretty for beautification, but my fear is that jspretty would mutilate the code.

**Comment by lenilsondc** [comment]
I mean, the system of using an external beautifier-analizer. I sure will have to create a VBScript parser similar to jspretty.js but I'm just worried in how to integrate it properly to prettydiff.

**Comment by prettydiff** [comment]
What I would do is to allow an option that expects a command line instruction in a child process using Node.js.  Once markuppretty has finished beautifying the rest of the document I would simply spawn a child process in node and pass in tokens of types `template`, `template_start`, and `template_end` to that child process.

I will start working on it this weekend and together we can figure out how to make this work efficiently and asynchronously for maximum speed and ease of use.

**Comment by prettydiff** [comment]
Could you provide an example of a command line instruction you might use to beautify ASP code with a VBScript application?

**Comment by lenilsondc** [comment]
Well the beautifier is quite simple but I'm afraid that this way the input arguments may cause some divergence with the current pattern.

````
vbsbeautify in:"input/to/file" out:"output/to/file" indentChar:"\t" indentLevel:2
````

**Comment by prettydiff** [comment]
The options format is identify to how Node.js takes arguments from the command line.  I am not worried about differences in the commands or actual options.  The challenges are as follows:

* **Deliver a string from the markuppretty library to the command line for evaluation.**
* Execute the third party beautifier.  This is actually easy, but demands some care when working with multiple items simultaneously.  You can execute command line instructions as child process in Node and just await the asynchronous callback.
* Take the output from the third party library as a string directly back into the markuppretty library to replace the ASP token.  I also suspect this won't be hard as the third party application should return its output as standard output to the command line which Node will intercept and return as an argument in the callback.

I just need to find an efficient way to deliver strings from markuppretty.js to the command line and everything else is just writing code.  I don't want to write anything to files during this process, because that would be very slow.  I pride my application on being fast down the millisecond.  I also cannot deliver a simple string to the command line with the code directly injected as a string fragment in the command, because I believe Windows will only tolerate commands of a certain length.

**Comment by lenilsondc** [comment]
Understood!
It seems to be a hard choice.
If you're talking about `child_process` I think the best shot is to create an IPC channel.
I'm not very familiar with `child_process` but from what I know, it also may cause performance issues which scales with the size of message sent to the child process.

I'll research a little more later.

**Comment by lenilsondc** [comment]
I've read about it and found out that sending messages to process will always be expensive in terms of performance. And also read that a better approach is sending the `responder` (socket)  wich in most of cases would be a more lightweight message that responds (streams) directly from the worker process.

Other approach would be write to file system, but I think it's not an option for this case.

References.

1. [sending a socket object](https://nodejs.org/api/child_process.html#child_process_example_sending_a_socket_object)

**Comment by prettydiff** [comment]
Thank you, reading now...

**Comment by prettydiff** [comment]
Basic code reorganization for asynchronous operations is complete.  Now the real work can begin.

**Comment by prettydiff** [comment]
This is proving challenging.  Something I did not consider is that the modularity of the various files is synchronous.  This is more than just reorganizing a library file to anticipate asynchronous operations, but also the relationship between the various files.  I need to rethink code organization on a larger scale to make this work.

**Comment by prettydiff** [comment]
This issue demands high scrutiny and effort than originally envisioned.  Moving to next milestone so that I can publish the current resolutions.

**Comment by prettydiff** [comment]
Effort so far is stored in branch external

**Comment by lenilsondc** [comment]
Sorry for the late response, I am working on the `vbs` beautifier and now I'm running for tests, optimizations and stuff. As soon as I can I will take a try on prettydiff's `external` branch.

For now it's called `vbspretty`, I don't know yet but it may change to `vbs6pretty` because last asp vbs engine was vbs 6. Whatever, I might keep vbspretty, it look cool.

Cheers 

:{D

**Comment by prettydiff** [comment]
I think there is a lot of work I still need to do on that branch.  Right now my attention is primarily upon solving a publication/distribution gap since I have abandoned NPM, but I can circle back to this when the need arises.

**Comment by lenilsondc** [comment]
Hello mate!
I am finally retaking the project of the asp beautifyer and taking a look in markuppretty.js I came up with the idea of instead using a children process to delegate the beautification of the templates, it could be done using a submode for the beautify mode, couldn't it?

So, let me depict properly what I'm talking about.

This submode of beautify mode should return an object with the necessary data to identify the templates that should be cared about, instead of a beautified output.

So then, when you are using pretidiff as a dependence library, you could ask for prettydiff to return that object and then with the proper data in hands you can apply the beautification for an especific language or whatever.

The below code was used for testing what could be done. Not solid enough but express the idea.

````javascript
if(options.should_return_object_instead_of_string){
    var markups = [];
    types.map(function(type, i) {

        if (type !== 'template') {
            return;
        }

        markups.push({
          output: output,
          attrs: attrs[i],
          begin: begin[i],
          daddy: daddy[i],
          jscom: jscom[i],
          level: level[i],
          linen: linen[i],
          lines: lines[i],
          presv: presv[i],
          token: token[i],
          types: types[i]
        });
    });

   return markup;
}
return output;
````

Of course `options.should_return_object_instead_of_string` requires a more fancier name. :{D

Anyway, fell free to question my point, it's just an idea that could make the solution more simpler (perhaps not) without affect the library speed when using the normal modes, only when is required this to be done.

**Comment by prettydiff** [comment]
> it could be done using a submode for the beautify mode, couldn't it?

Yes, it could, but to be efficient it should be asynchronous.  The challenge there is that I would have to rewrite the code to allow asynchronous operations.  Currently the two API files are written entirely with asynchronous execution in mind, but the parsers and other libraries are not.  The potential for asynchronous execution is that I can dispatch a large block of code, such as PHP or ASP code, to their respective beautifier while Pretty Diff's parsers continue forward with the rest of the document.  The challenge there is that I would have to find a way to halt the program until all the pieces have returned into the correct places and do this in a way that works on multiple files simultaneously.

**Comment by prettydiff** [comment]
Also, while I still not yet supporting VBScript or PHP I am about to release support for TypeScript, Java, and C#.  I found that when I extended my beautifier to support TypeScript I can support the other two languages almost for free.

**Comment by prettydiff** [comment]
This is problem is solved in the 3.0.0 branch.  The next version of Pretty Diff offloads all parsing tasks to a tool called [parse-framework](https://github.com/Unibeautify/parse-framework/).  Any compliant parser may be used for any grammar in that framework, but they need to be written.

**Comment by prettydiff** [comment]
This issue here was my tada moment.  This is the inception point from which all architecture has since happened.

For HTML embedded PHP specifically here is what needs to happen now with the new parser:

1. Write a PHP lexer using the universal parser I have started at Unibeatify/parse-framework https://github.com/unibeautify/parse-framework
1. In the parser modify the markup lexer to send PHP segments to the PHP lexer.
1. Beautify all parse records of `lexer=php` with whatever beautifier you wish
1. Send the modified document to Pretty Diff to beautify the rest
